KalturaWowzaServer/src/main/java/com/kaltura/media_server/modules/LiveStreamSettingsModule.java

```diff
@@ -0,0 +1,340 @@
+					properties.setProperty(Constants.AUDIO_BASE_SYSTEM_TIME, baseSystemTime);
+					properties.setProperty(Constants.AUDIO_BASE_PTS, baseInPTS);
+					properties.setProperty(Constants.AUDIO_LAST_IN_PTS, baseInPTS);
+
+				} else if (thisPacket.isVideo()) {
+					properties.setProperty(Constants.VIDEO_BASE_SYSTEM_TIME, baseSystemTime);
+					properties.setProperty(Constants.VIDEO_BASE_PTS, baseInPTS);
+					properties.setProperty(Constants.VIDEO_LAST_IN_PTS, baseInPTS);
+				} else {
+					properties.setProperty(Constants.DATA_BASE_SYSTEM_TIME, baseSystemTime);
+					properties.setProperty(Constants.DATA_BASE_PTS, baseInPTS);
+					properties.setProperty(Constants.DATA_LAST_IN_PTS, baseInPTS);
+				}
+			} else {
+				if (thisPacket.isAudio()) {
+					properties.setProperty(Constants.AUDIO_LAST_IN_PTS, inPTS);
+				} else if (thisPacket.isVideo()) {
+					properties.setProperty(Constants.VIDEO_LAST_IN_PTS, inPTS);
+				} else {
+					properties.setProperty(Constants.DATA_LAST_IN_PTS, inPTS);
+				}
+			}
+
+			//=================================================================
+			// calc & update output PTS value
+			//=================================================================
+			long correction = baseSystemTime - baseInPTS;
+			long outPTS = inPTS + correction;
+
+			thisPacket.setAbsTimecode(outPTS);
+
+			if (firstPacket) {
+				logger.debug("(" + streamName + ") [" + streamType + "] first PTS updated to [" + outPTS + "] PTS was [" + inPTS + "] basePTS [" + baseInPTS + "] baseSystemTime [" + baseSystemTime + "]");
+			} else if (absPTSTimeCodeDiff > maxAllowedPTSDriftMillisec) {
+				logger.warn("(" + streamName + ") [" + streamType + "] PTS diff[" + inPTSDiff + "] > threshold [" + maxAllowedPTSDriftMillisec + "] last PTS [" + lastInPTS + "] current PTS [" + inPTS + "] basePTS [" + baseInPTS + "] baseSystemTime [" + baseSystemTime + "]");
+			} else {
+				logger.debug("(" + streamName + ") [" + streamType + "] updated PTS [" + outPTS + "] in PTS [" + inPTS + "]");
+			}
+		}
+	}
+
+	public long[] getLiveEntryBaseSystemTime(String entryId, long baseSystemTime, long basePTSTimeCode) {
+
+		long[] refData = {baseSystemTime, basePTSTimeCode};
+
+		synchronized (this.mapLiveEntryToBaseSystemTime) {
+			try {
+				if (!this.mapLiveEntryToBaseSystemTime.containsKey(entryId)) {
+					this.mapLiveEntryToBaseSystemTime.put(entryId, refData);
+				} else {
+					refData = this.mapLiveEntryToBaseSystemTime.get(entryId);
+				}
+			} catch (Exception e) {
+				if (Boolean.valueOf(entryId)) {
+					logger.error("(" + entryId + ") fail to get base timestamp for live entry." + e.toString());
+				} else {
+					logger.error("( undefined entryId ) could not get base timestamp, entryId is invalid. " + e.toString());
+				}
+			}
+		}
+
+		return refData;
+	}
+
+	public long[] setLiveEntryBaseSystemTime(String entryId, long baseSystemTime, long basePTS) {
+
+		long[] globalData = this.getLiveEntryBaseSystemTime(entryId, baseSystemTime, basePTS);
```

Comment:
there is a race between getLiveEntryBaseSystemTime and  second lock, it should be on method with one lock